File: processing/syntactic/expressions/block_expr.py
from __future__ import annotations

from inspect import getmembers
from typing import List, Collection, Union, Optional, Type, TypeVar

from aenum import Flag, auto

import specification as spec
from errors.blame import BlameType, BlameSeverity
from processing.codegen.code_builder import CodeBuilder
from processing.lexical.tokens.token_type import TokenType
from processing.syntactic.expressions.definitions.var_def import VarDef
from processing.syntactic.expressions.expr import Expr, child_property
from processing.syntactic.expressions.groups import DefinitionExpression, VarTargetExpression
from processing.syntactic.parsing import parse_any


class BlockType(Flag):
    default = auto()
    ast = auto()
    named = auto()
    loop = auto()
    anon_fn = auto()


class BlockExpr(Expr):
    """ block:
        (':' expr)
        | ([':'] '{' expr* '}')
        | ([':'] NEWLINE INDENT expr+ OUTDENT);
    """

    @child_property
    def items(self) -> List[Expr]:
        pass

    def __init__(self, parent: Expr = None, items: List[Expr] = None):
        super().__init__(parent)
        self.items = items

    def is_defined(self, name: Union[str, VarTargetExpression]) -> bool:
        from processing.syntactic.expressions.ast import Ast

        if not isinstance(self, Ast) and self.get_parent_of_type(BlockExpr).is_defined(name):
            return True
        defs = self.get_all_defs()
        if isinstance(name, Collection):
            for n in name:
                if str(n) in defs:
                    return True
            return False
        else:
            return str(name) in defs

    def get_def_by_name(self, name: Union[str, VarTargetExpression]) -> Optional[Expr]:
        from processing.syntactic.expressions.ast import Ast

        if not isinstance(self, Ast):
            e = self.get_parent_of_type(BlockExpr).get_def_by_name(name)
            if e is not None:
                return e
        defs = self.get_all_defs()
        if isinstance(name, Collection):
            for n in name:
                s = str(n)
                if s in defs:
                    return defs[s]
            return None
        else:
            return defs.get(str(name))

    def get_all_defs(self):
        from processing.syntactic.expressions.definitions.func_def import FuncDef

        defs = {str(i.name): i for i in self.items if isinstance(i, DefinitionExpression)}
        parent_fn = self.get_parent_of_type(FuncDef)
        if parent_fn is not None:
            defs[str(parent_fn.name)] = parent_fn.name
            defs.update({str(p): p for p in parent_fn.parameters})
        return defs

    T = TypeVar('T', bound = Expr)

    def find_items_of_type(self, typ: Type[T], outs: List[Expr] = None) -> List[T]:
        outs = outs or []
        for i in self.items:
            if isinstance(i, typ):
                outs.append(i)
            elif isinstance(i, BlockExpr):
                i.find_items_of_type(typ, outs)
                child_blocks = getmembers(i, lambda o: isinstance(o, BlockExpr))
                for b in child_blocks:
                    b.find_items_of_type(typ, outs)
        return outs

    def parse(self, block_type: BlockType) -> BlockExpr:
        if block_type == BlockType.ast:
            terminator = TokenType.end
        elif not self.stream.peek_is(*spec.block_start_types):
            return self
        else:
            terminator = self.parse_start()
        if BlockType.anon_fn in block_type:
            if terminator == TokenType.outdent:
                self.source.blame(BlameType.lambda_cannot_have_indented_body, self)
            elif terminator == TokenType.newline:
                self.items.append(parse_any(self))
                return self
        while not self.stream.maybe_eat(terminator) \
                and not self.stream.peek_is(TokenType.end) \
                and not (terminator == TokenType.newline and self.stream.token.of_type(TokenType.newline)):
            self.items.append(parse_any(self))
        return self

    def parse_start(self) -> TokenType:
        has_colon = self.stream.maybe_eat(TokenType.colon)
        block_start = self.stream.token
        has_newline = self.stream.maybe_eat(TokenType.newline) if has_colon \
            else block_start.of_type(TokenType.newline)
        if self.stream.maybe_eat(TokenType.open_brace):
            if has_colon:
                self.source.blame(BlameType.redundant_colon_with_braces, self.stream.peek)
            return TokenType.close_brace
        if self.stream.maybe_eat(TokenType.indent):
            return TokenType.outdent
        if has_newline:
            self.source.blame(BlameType.expected_block_declaration, self.stream.peek)
            return TokenType.newline
        if not has_colon:
            self.source.blame("expected ':'", self.stream.peek, BlameSeverity.error)
            return TokenType.newline
        return TokenType.newline

    def to_axion(self, c: CodeBuilder):
        c.indent()
        c += self.items
        c.outdent()

    def to_csharp(self, c: CodeBuilder):
        from processing.syntactic.expressions.conditional_expr import ConditionalExpr
        from processing.syntactic.expressions.while_expr import WhileExpr
        c.write_line()
        c.write_line('{')
        c.indent()
        for item in self.items:
            c += item
            if isinstance(item, (DefinitionExpression, ConditionalExpr, WhileExpr)) \
                    and not isinstance(item, VarDef):
                c.write_line()
            else:
                c.write_line(';')
        c.outdent()
        c += '}'

    def to_python(self, c: CodeBuilder):
        c += ':'
        c.indent()
        c.write_line()
        for item in self.items:
            c += item
            c.write_line()
        c.outdent()

File: Axion.Core/Processing/Syntactic/Expressions/BlockExpr.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using Axion.Core.Processing.CodeGen;
using Axion.Core.Processing.Errors;
using Axion.Core.Processing.Lexical.Tokens;
using Axion.Core.Processing.Syntactic.Expressions.Definitions;
using Axion.Core.Specification;
using static Axion.Core.Processing.Lexical.Tokens.TokenType;

namespace Axion.Core.Processing.Syntactic.Expressions {
    /// <summary>
    ///     <c>
    ///         block:
    ///             (':' expr)
    ///             | ([':'] '{' expr* '}')
    ///             | ([':'] NEWLINE INDENT expr+ OUTDENT);
    ///     </c>
    /// </summary>
    public class BlockExpr : Expr {
        private NodeList<Expr> items;

        public NodeList<Expr> Items {
            get => items;
            protected set => SetNode(ref items, value);
        }

        internal BlockExpr(
            Expr          parent = null,
            params Expr[] items
        ) : base(parent) {
            Items = new NodeList<Expr>(this, items);

            if (Items.Count != 0) {
                MarkPosition(Items.First, Items.Last);
            }
        }

        protected BlockExpr() { }

        public string CreateUniqueId(string formattedId) {
            var    i  = 0;
            string id = string.Format(formattedId, i);
            while (IsDefined(id)) {
                i++;
                id = string.Format(formattedId, i);
            }

            return id;
        }

        public bool IsDefined(string name) {
            return GetDefByName(name) != null;
        }

        public IDefinitionExpr GetDefByName(string name) {
            if (!(this is Ast)) {
                IDefinitionExpr e = GetParentOfType<BlockExpr>().GetDefByName(name);
                if (e != null) {
                    return e;
                }
            }

            IDefinitionExpr[] defs = GetAllDefs();
            return defs.FirstOrDefault(def => def.Name.ToString() == name);
        }

        public IDefinitionExpr[] GetAllDefs() {
            List<IDefinitionExpr> defs = Items.OfType<IDefinitionExpr>().ToList();

            // Add parameters of function if inside it.
            var parentFn = GetParentOfType<FunctionDef>();
            if (parentFn != null) {
                defs.AddRange(parentFn.Parameters);
            }

            return defs.ToArray();
        }

        public List<(T item, BlockExpr itemParentBlock, int itemIndex)> FindItemsOfType<T>(
            List<(T item, BlockExpr itemParentBlock, int itemIndex)> _outs = null
        ) where T : Expr {
            _outs = _outs ?? new List<(T item, BlockExpr itemParentBlock, int itemIndex)>();
            for (var i = 0; i < Items.Count; i++) {
                Expr item = Items[i];
                if (item is T expr) {
                    _outs.Add((expr, this, i));
                }
                else {
                    IEnumerable<PropertyInfo> childBlockProps = item.GetType().GetProperties().Where(
                        p => p.PropertyType == typeof(BlockExpr)
                          && p.Name         != nameof(Parent)
                    );
                    foreach (PropertyInfo blockProp in childBlockProps) {
                        var b = (BlockExpr) blockProp.GetValue(item);
                        b?.FindItemsOfType(_outs);
                    }
                }
            }

            return _outs;
        }

        public (BlockExpr itemParentBlock, int itemIndex) IndexOf<T>(T expression) where T : Expr {
            for (var i = 0; i < Items.Count; i++) {
                Expr item = Items[i];
                if (item == expression) {
                    return (this, i);
                }

                IEnumerable<PropertyInfo> childBlockProps = item.GetType().GetProperties().Where(
                    p => p.PropertyType == typeof(BlockExpr)
                      && p.Name         != nameof(Parent)
                );
                foreach (PropertyInfo blockProp in childBlockProps) {
                    var                                         b   = (BlockExpr) blockProp.GetValue(item);
                    (BlockExpr itemParentBlock, int itemIndex)? idx = b?.IndexOf(expression);
                    if (idx != null && idx != (null, -1)) {
                        return ((BlockExpr itemParentBlock, int itemIndex)) idx;
                    }
                }
            }

            return (null, -1);
        }

        public BlockExpr Parse(BlockType blockType = BlockType.Default) {
            if (!Stream.PeekIs(Spec.BlockStartMarks)) {
                return this;
            }

            SetSpan(() => {
                TokenType terminator = ParseStart(this);

                if (terminator == Outdent && blockType.HasFlag(BlockType.Lambda)) {
                    LangException.Report(BlameType.IndentationBasedBlockNotAllowed, this);
                }

                if (terminator == Newline) {
                    Items.Add(Parsing.ParseAny(this));
                }
                else {
                    while (!Stream.MaybeEat(terminator)
                        && !Stream.PeekIs(TokenType.End)
                        && !(terminator == Newline && Stream.Token.Is(Newline))) {
                        Items.Add(Parsing.ParseAny(this));
                    }
                }
            });
            return this;
        }

        /// <summary>
        ///     Starts parsing the statement's block,
        ///     returns terminator what can be used to parse block end.
        /// </summary>
        internal static TokenType ParseStart(Expr parent) {
            // colon
            bool  hasColon   = parent.Stream.MaybeEat(Colon);
            Token blockStart = parent.Stream.Token;

            // newline
            bool hasNewline = hasColon
                ? parent.Stream.MaybeEat(Newline)
                : blockStart.Is(Newline);

            // '{'
            if (parent.Stream.MaybeEat(OpenBrace)) {
                if (hasColon) { // ':' '{'
                    LangException.Report(BlameType.RedundantColonWithBraces, blockStart);
                }

                return CloseBrace;
            }

            // indent
            if (parent.Stream.MaybeEat(Indent)) {
                return Outdent;
            }

            if (hasNewline) {
                // newline followed by not indent or '{'
                LangException.Report(BlameType.ExpectedBlockDeclaration, parent.Stream.Peek);
            }
            // exactly a 1-line block
            else if (!hasColon) {
                // one line block must have a colon
                LangException.ReportUnexpectedSyntax(Colon, parent.Stream.Peek);
            }

            return Newline;
        }

        public override void ToAxion(CodeWriter c) {
            bool inAnonFn = Parent is FunctionDef fn && fn.Name == null;
            if (inAnonFn) {
                c.WriteLine(" {");
            }

            c.Indent++;
            c.MaybeWriteLine();
            c.AddJoin("", Items, true);
            c.MaybeWriteLine();
            c.Indent--;
            
            if (inAnonFn) {
                c.Write("}");
            }
        }

        public override void ToCSharp(CodeWriter c) {
            c.WriteLine();
            c.WriteLine("{");
            c.Indent++;
            foreach (Expr item in Items) {
                c.Write(item);
                if (!(item is IDefinitionExpr
                   || item is ConditionalExpr
                   || item is WhileExpr) || item is VarDef) {
                    c.Write(";");
                }

                c.MaybeWriteLine();
            }

            c.Indent--;
            c.WriteLine("}");
        }

        public override void ToPython(CodeWriter c) {
            c.Write(":");
            c.Indent++;
            c.WriteLine();
            c.AddJoin("", Items, true);
            c.Indent--;
            c.MaybeWriteLine();
        }
    }

    [Flags]
    public enum BlockType {
        Default,
        Named,
        Loop,
        Lambda
    }
}